src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs

```diff
@@ -28,642 +37,878 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
+                }
+                else
+                {
+                    Assert.Equal(
+    @"Delta: Gamma: Test G
 Delta: Epsilon: Test E
 ",
-                    actual);
-            }
+                        actual);
+                }
+            });
         }
 
         [Fact]
         public void AssemblyLoading_MultipleVersions_NoExactMatch()
         {
-            StringBuilder sb = new StringBuilder();
+            Run(static (DefaultAnalyzerAssemblyLoader loader, AssemblyLoadTestFixture testFixture) =>
+            {
+                StringBuilder sb = new StringBuilder();
 
-            var loader = new DefaultAnalyzerAssemblyLoader();
-            loader.AddDependencyLocation(_testFixture.Delta1.Path);
-            loader.AddDependencyLocation(_testFixture.Epsilon.Path);
-            loader.AddDependencyLocation(_testFixture.Delta3.Path);
+                loader.AddDependencyLocation(testFixture.Delta1.Path);
+                loader.AddDependencyLocation(testFixture.Epsilon.Path);
+                loader.AddDependencyLocation(testFixture.Delta3.Path);
 
-            Assembly epsilon = loader.LoadFromPath(_testFixture.Epsilon.Path);
-            var e = epsilon.CreateInstance("Epsilon.E")!;
-            e.GetType().GetMethod("Write")!.Invoke(e, new object[] { sb, "Test E" });
+                Assembly epsilon = loader.LoadFromPath(testFixture.Epsilon.Path);
+                var e = epsilon.CreateInstance("Epsilon.E")!;
+                e.GetType().GetMethod("Write")!.Invoke(e, new object[] { sb, "Test E" });
 
-#if NETCOREAPP
-            var alcs = DefaultAnalyzerAssemblyLoader.TestAccessor.GetOrderedLoadContexts(loader);
-            Assert.Equal(1, alcs.Length);
-
-            VerifyAssemblies(
-                alcs[0].Assemblies,
-                ("Delta", "3.0.0.0", _testFixture.Delta3.Path),
-                ("Epsilon", "0.0.0.0", _testFixture.Epsilon.Path));
-#endif
-
-            var actual = sb.ToString();
-            if (ExecutionConditionUtil.IsCoreClr)
-            {
-                Assert.Equal(
+                var actual = sb.ToString();
+                if (ExecutionConditionUtil.IsCoreClr)
+                {
+                    // In .NET Core we have _full_ control over assembly loading and can prevent implicit
+                    // loads from probing paths. That means we can avoid implicitly loading the Delta v2 
+                    // next to Epsilon
+                    VerifyDependencyAssemblies(
+                        loader,
+                        testFixture.Delta3.Path,
+                        testFixture.Epsilon.Path);
+                    Assert.Equal(
 @"Delta.3: Epsilon: Test E
 ",
                     actual);
-            }
-            else
-            {
-                Assert.Equal(
-@"Delta: Epsilon: Test E
+                }
+                else
+                {
+                    // The Epsilon.dll has Delta.dll (v2) next to it in the directory. The .NET Framework 
```

Comment:
Could you explain the behavior change here a bit more? I am afraid that maybe before this PR the behavior depended on the order of execution of tests 😓

Reply:
When you load a DLL via `LoadFrom` the path of that DLL becomes a part of the probing path for that DLL (but not others). In this case the director layout is as follows:

- Version2
    - Epsilon.dll 
    - Delta.dll (version 2) 

The compile loads Epsilon.dll into the _LoadFrom_ context via the `loader.LoadFromPath(testFixture.Epsilon.Path)` call above. That puts the Version2 directory on the private probing path for Epsilon.dll. When the time comes to resolve Delta.dll the runtime notices it's not available in the _LoadFrom_ or _Load_ context so it attempts to load it. Because Delta.dll is on the private probing path this load happens implicitly and the DLL is resolved.

There is no call to `AppDomain.ResolveAssembly` here so no chance for us to intercept this load. It is completely outside our control. Unlike in .NET Core where every load is controlled by an `AssemblyLoadContext` and we can force it to only pick registered dependencies. 

I do suspect this test was working due to prior test pollution before. Prior tests would've put a different version of Delta.dll into _LoadFrom_ and that would've satisfied the implicit load.